plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php

```diff
@@ -351,27 +393,124 @@ public function cleanup_synchronization_state() {
 			$this->order_cache_controller->maybe_restore_orders_cache_usage();
 		}
 	}
 
+	/**
+	 * Take a batch of order ids pending synchronization and process those that were deleted, ignoring the others
+	 * (which will be orders that were created or modified) and returning the ids of the orders actually processed.
+	 *
+	 * @param array $batch Array of ids of order pending synchronization.
+	 * @param bool  $custom_orders_table_is_authoritative True if the custom orders table is currently authoritative.
+	 * @return array Order ids that have been actually processed.
+	 */
+	private function process_deleted_orders( array $batch, bool $custom_orders_table_is_authoritative ): array {
+		global $wpdb;
+
+		$deleted_from_table_name = $custom_orders_table_is_authoritative ? $this->data_store::get_orders_table_name() : $wpdb->posts;
+		$data_store_for_deletion =
+			$custom_orders_table_is_authoritative ?
+			new \WC_Order_Data_Store_CPT() :
+			wc_get_container()->get( OrdersTableDataStore::class );
+
+		$order_ids_as_sql_list = '(' . implode( ',', $batch ) . ')';
+
+		$deleted_order_ids  = array();
+		$meta_ids_to_delete = array();
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$deletion_data = $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT id, order_id FROM {$wpdb->prefix}wc_orders_meta WHERE meta_key=%s AND meta_value=%s AND order_id IN $order_ids_as_sql_list ORDER BY order_id DESC",
+				'deleted_from',
+				$deleted_from_table_name
+			),
+			ARRAY_A
+		);
+		// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+
+		if ( empty( $deletion_data ) ) {
+			return array();
+		}
+
+		foreach ( $deletion_data as $item ) {
+			$meta_id  = $item['id'];
+			$order_id = $item['order_id'];
+
+			if ( isset( $deleted_order_ids[ $order_id ] ) ) {
+				$meta_ids_to_delete[] = $meta_id;
+				continue;
+			}
+
+			if ( ! $data_store_for_deletion->order_exists( $order_id ) ) {
+				$this->error_logger->warning( "Order {$order_id} doesn't exist in the backup table, thus it can't be deleted" );
+				$deleted_order_ids[]  = $order_id;
+				$meta_ids_to_delete[] = $meta_id;
+				continue;
+			}
+
+			try {
+				$order = new \WC_Order();
+
+				// We need to turn parent order verification off to avoid an exception for child orders (e.g. refunds).
+				// The 'read' method in the order class invokes 'set_parent_id', which by default verifies
+				// if the order the parent id refers to actually exists, and throws an exception if not.
+				// To check if the order exists 'wc_get_order' is used, but this function searches the order
+				// using the authoritative data store, and at this point the order doesn't exist anymore in the
+				// authoritative table ('read' below is being invoked in the backup data store).
+				$order->set_verify_parent_id( false );
+
+				$order->set_id( $order_id );
+				$data_store_for_deletion->read( $order );
+
+				$data_store_for_deletion->delete( $order, array( 'force_delete' => true ) );
```

Comment:
this will fire wc hooks again, perhaps we need some kind of suppress filter mechanism here?

Reply:
Maybe we could add a `suppress_firing_hooks` in the passed arguments array, and look for it in the `delete` method of both the CPT and the new data store class?

Reply:
I think naming it `$suppress_filters` would keep it consistent with [existing WP_Query nomenclature](https://github.com/WordPress/wordpress-develop/blob/6.2/src/wp-includes/class-wp-query.php#L759). But otherwise yeah, that makes sense.

Reply:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/406e2acba736357e1daeaa26631adc910e388ca0 for `OrdersTableDataStore`, and in https://github.com/woocommerce/woocommerce/pull/37050/commits/77af8532c62ea25b84eb93e55770150a7fea54e9 for `Abstract_WC_Order_Data_Store_CPT`.

Additionally, in https://github.com/woocommerce/woocommerce/pull/37050/commits/77af8532c62ea25b84eb93e55770150a7fea54e9 the  `Abstract_WC_Order_Data_Store_CPT` method has been modified to fire `woocommerce_before_delete/trash_order` actions, for symmetry with `OrdersTableDataStore`.